src/client/opamSolution.ml

```diff
@@ -474,6 +478,79 @@ let parallel_apply t
     else action_graph
   in
 
+  (* factorise fetches: creates [`Fetches] nodes that download once the archive
+     and copies its content in source directories. Handles only http and rsync
+     backend with at least a checksum. *)
+  let action_graph =
+    let cache_dir = OpamRepositoryPath.download_cache t.switch_global.root in
+    let packages =
+      let urls, hash_nvs =
+        OpamPackage.Set.fold (fun nv (urls, hashes as acc) ->
+            match OpamSwitchState.url t nv with
+            | Some url ->
+              let checksums = OpamFile.URL.checksum url in
+              let url = OpamFile.URL.url url in
+              (match url.OpamUrl.backend, checksums with
+               | (`http | `rsync ) , _::_
+                 when OpamFilename.is_archive
+                     (OpamFilename.of_string url.OpamUrl.path) ->
+                 OpamUrl.Map.update url
+                   (fun hset -> OpamHash.Set.(union hset (of_list checksums)))
+                   OpamHash.Set.empty urls,
+                 List.fold_left (fun hashes h ->
+                     OpamHash.Map.update h (OpamPackage.Set.add nv)
+                       OpamPackage.Set.empty hashes)
+                   hashes checksums;
+               | _ -> acc)
+            | None -> acc)
+          sources_needed (OpamUrl.Map.empty, OpamHash.Map.empty)
+      in
+      OpamUrl.Map.fold (fun _url hashes packages ->
```

Comment:
Folding by url is important here: otherwise there could be an attack using a collision on a weak hash (md5) on a different package/url. It would be detected later on since we check all hashes upon download, but would still deny installation…

Reply:
outdated